lib/ansible/modules/cloud/amazon/ec2_group.py

```diff
@@ -288,66 +288,156 @@
+    except:
+        raise Exception("Failed on rule %s" % str(rule))
+    if rule.target_type == 'ipv4':
+        perm['IpRanges'] = [{
+            'CidrIp': rule.target,
+        }]
+        if rule.description:
+            perm['IpRanges'][0]['Description'] = rule.description
+    elif rule.target_type == 'ipv6':
+        perm['Ipv6Ranges'] = [{
+            'CidrIpv6': rule.target,
+        }]
+        if rule.description:
+            perm['Ipv6Ranges'][0]['Description'] = rule.description
+    elif rule.target_type == 'group':
+        if isinstance(rule.target, tuple):
+            perm['UserIdGroupPairs'] = [{
+                'OwnerId': rule.target[0],
+                'GroupId': rule.target[1],
+                'GroupName': rule.target[2],
+            }]
+            raise NotImplementedError("Whoopsie, no foreign security group serialization yet. Target {0}".format(rule.target))
+        else:
+            perm['UserIdGroupPairs'] = [{
+                'GroupId': rule.target
+            }]
+        if rule.description:
+            perm['UserIdGroupPairs'][0]['Description'] = rule.description
+    elif rule.target_type == 'ip_prefix':
+        perm['PrefixListIds'] = [{
+            'PrefixListId': rule.target,
+        }]
+        if rule.description:
+            perm['PrefixListIds'][0]['Description'] = rule.description
+    elif rule.target_type not in valid_targets:
+        raise ValueError('Invalid target type for rule {0}'.format(rule))
+    return fix_port_and_protocol(perm)
+
+
+def rule_from_group_permission(perm):
+    def ports_from_permission(p):
+        if 'FromPort' not in p and 'ToPort' not in p:
+            return (None, None)
+        return (int(perm['FromPort']), int(perm['ToPort']))
+
+    try:
+        # outputs a rule tuple
+        for target_key, target_subkey, target_type in [
+            ('IpRanges', 'CidrIp', 'ipv4'),
+            ('Ipv6Ranges', 'CidrIpv6', 'ipv6'),
+            ('PrefixListIds', 'PrefixListId', 'ip_prefix'),
+        ]:
+            if target_key not in perm:
+                continue
+            for r in perm[target_key]:
+                # there may be several IP ranges here, which is ok
+                yield Rule(
+                    ports_from_permission(perm),
+                    perm['IpProtocol'],
+                    r[target_subkey],
+                    target_type,
+                    r.get('Description')
+                )
+        if 'UserIdGroupPairs' in perm and perm['UserIdGroupPairs']:
+            # security group targets are special, handle separately
+            # TODO this doesn't handle cross-account groups yet
+            for pair in perm['UserIdGroupPairs']:
+                yield Rule(
+                    ports_from_permission(perm),
+                    perm['IpProtocol'],
+                    pair['GroupId'],
+                    'group',
+                    pair.get('Description')
+                )
+    except Exception as e:
```

Comment:
Just making a note that we should probably restrict this to the exceptions that will likely occur and call fail_json instead of raising an Exception again.